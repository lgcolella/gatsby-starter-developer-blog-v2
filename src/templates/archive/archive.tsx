/* Vendor imports */
import React from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'gatsby'
/* App imports */
import { AvailableTag, ChildImageSharpFluid } from '../../types'
import Layout from '../../components/layout'
import SEO from '../../components/seo'
import PostList from '../../components/post-list'
import ArchivePagination from '../../components/archive-pagination'
import Config from '../../../config'
import Lang from '../../i18n/default'

const Archive = ({ data, pageContext }: Props) => {
  const { archivePage, lastArchivePage } = pageContext
  const prevPage = archivePage > 1 ? archivePage - 1 : undefined
  const nextPage = archivePage < lastArchivePage ? archivePage + 1 : undefined

  return (
    <Layout title={Lang.templates.archive.title}>
      <SEO
        title={`${Lang.templates.archive.pageTitle} ${archivePage}`}
        description={Lang.templates.archive.description}
        path={Config.pages.archive}
      />
      <PostList posts={data.allMarkdownRemark.edges} />
      <ArchivePagination prevPage={prevPage} nextPage={nextPage} />
    </Layout>
  )
}

Archive.propTypes = {
  data: PropTypes.shape({
    allMarkdownRemark: PropTypes.shape({
      edges: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired,
    }).isRequired,
  }).isRequired,
  pageContext: PropTypes.shape({
    postIds: PropTypes.arrayOf(PropTypes.string).isRequired,
    archivePage: PropTypes.number.isRequired,
    lastArchivePage: PropTypes.number.isRequired,
  }).isRequired,
}

export interface Props {
  data: {
    allMarkdownRemark: {
      edges: Array<{
        node: {
          frontmatter: {
            title: string
            path: string
            tags: AvailableTag[]
            date: string
            excerpt: string
            cover: {
              childImageSharp: ChildImageSharpFluid
            }
          }
        }
      }>
    }
  }
  pageContext: {
    postIds: String[]
    archivePage: number
    lastArchivePage: number
  }
}

export const query = graphql`
  query($postIds: [String!]) {
    allMarkdownRemark(
      filter: {
        id: { in: $postIds }
        fileAbsolutePath: { regex: "/index.md$/" }
      }
      sort: { fields: [frontmatter___date], order: DESC }
    ) {
      edges {
        node {
          frontmatter {
            title
            path
            tags
            date(formatString: "YYYY-MM-DD")
            excerpt
            cover {
              childImageSharp {
                fluid(maxWidth: 600) {
                  ...GatsbyImageSharpFluid_tracedSVG
                }
              }
            }
          }
        }
      }
    }
  }
`

export default Archive
