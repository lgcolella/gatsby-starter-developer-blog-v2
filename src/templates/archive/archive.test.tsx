/* Vendor imports */
import React from 'react'
import { shallow } from 'enzyme'
/* App imports */
import Archive from './index'
import { Props } from './archive'
import ArchivePagination from '../../components/archive-pagination'
import { ImageSharpFluid } from '../../../tests/mocks'

describe('Archive template', () => {
  const postLists: Props['data'] = {
    allMarkdownRemark: {
      edges: [
        // Article #1
        {
          node: {
            frontmatter: {
              title: 'Title of article #1',
              path: 'path-to-the-article-1',
              tags: ['laravel', 'javascript', 'angular'],
              date: '2019-05-14',
              excerpt: 'an excertp for the article #1',
              cover: {
                childImageSharp: ImageSharpFluid,
              },
            },
          },
        },
        // Article #2
        {
          node: {
            frontmatter: {
              title: 'Title of article #2',
              path: 'path-to-the-article-2',
              tags: ['nodejs', 'angular'],
              date: '2018-12-15',
              excerpt: 'an excertp for the article #2',
              cover: {
                childImageSharp: ImageSharpFluid,
              },
            },
          },
        },
      ],
    },
  }

  test('matches snapshot', () => {
    let wrapper = shallow(
      <Archive
        data={postLists}
        pageContext={{
          archivePage: 3,
          lastArchivePage: 12,
          postIds: ['1', '2'],
        }}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  test('properly handles the pagination', () => {
    let wrapper = shallow(
      <Archive
        data={postLists}
        pageContext={{
          archivePage: 1,
          lastArchivePage: 1,
          postIds: ['1', '2'],
        }}
      />
    )

    // The archive has only one page
    expect(wrapper.find(ArchivePagination).prop('prevPage')).toBe(undefined)
    expect(wrapper.find(ArchivePagination).prop('nextPage')).toBe(undefined)
    // The archive has several pages and we are in the first one
    wrapper.setProps({
      pageContext: { archivePage: 1, lastArchivePage: 7, postIds: ['1', '2'] },
    })
    expect(wrapper.find(ArchivePagination).prop('prevPage')).toBe(undefined)
    expect(wrapper.find(ArchivePagination).prop('nextPage')).toBe(2)
    // The archive has several pages and we are in the middle
    wrapper.setProps({
      pageContext: { archivePage: 3, lastArchivePage: 7, postIds: ['1', '2'] },
    })
    expect(wrapper.find(ArchivePagination).prop('prevPage')).toBe(2)
    expect(wrapper.find(ArchivePagination).prop('nextPage')).toBe(4)
    // The archive has several pages and we are in the last one
    wrapper.setProps({
      pageContext: { archivePage: 7, lastArchivePage: 7, postIds: ['1', '2'] },
    })
    expect(wrapper.find(ArchivePagination).prop('prevPage')).toBe(6)
    expect(wrapper.find(ArchivePagination).prop('nextPage')).toBe(undefined)
  })
})
