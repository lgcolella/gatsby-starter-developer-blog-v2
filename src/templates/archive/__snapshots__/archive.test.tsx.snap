// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Archive template matches snapshot 1`] = `
<Layout
  className=""
  showHeader={true}
  title="Archive"
>
  <SEO
    description="The old posts of blog"
    path="archive"
    title="Archive | Page 3"
  />
  <PostList
    contentClassName=""
    posts={
      Array [
        Object {
          "node": Object {
            "frontmatter": Object {
              "cover": Object {
                "childImageSharp": Object {
                  "fluid": Object {
                    "aspectRatio": 1.7,
                    "sizes": "(max-width: 600px) 100vw, 600px",
                    "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                    "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                  },
                },
              },
              "date": "2019-05-14",
              "excerpt": "an excertp for the article #1",
              "path": "path-to-the-article-1",
              "tags": Array [
                "laravel",
                "javascript",
                "angular",
              ],
              "title": "Title of article #1",
            },
          },
        },
        Object {
          "node": Object {
            "frontmatter": Object {
              "cover": Object {
                "childImageSharp": Object {
                  "fluid": Object {
                    "aspectRatio": 1.7,
                    "sizes": "(max-width: 600px) 100vw, 600px",
                    "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                    "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
                  },
                },
              },
              "date": "2018-12-15",
              "excerpt": "an excertp for the article #2",
              "path": "path-to-the-article-2",
              "tags": Array [
                "nodejs",
                "angular",
              ],
              "title": "Title of article #2",
            },
          },
        },
      ]
    }
  />
  <ArchivePagination
    nextPage={4}
    prevPage={2}
  />
</Layout>
`;
